### PR TITLE
Documentation+AK: Update smart pointer documentation and make small improvements

### PR DESCRIPTION
**Pull request by BertalanD**
This PR documents the changes to memory allocation and updates to smart pointers introduced in #8150.

Also, dereference operators of smart pointers have been marked RETURNS_NONNULL, so that compilers can avoid null checks.

`VERIFY()` was changed in the kernel to mark the failure case as unlikely. This will reduce the cost of using `Nonnull{Own,Ref}Ptr` even with all checks enabled.

Construction of `NonnullRefPtr<T>` has been tidied up in a couple of places in the kernel.

### CODE REVIEW

File: Kernel/Assertions.h [hunk 1]
```diff
@@ -6,11 +6,13 @@
...
 #pragma once
 
+#include <AK/Platform.h>
+
 #define __STRINGIFY_HELPER(x) #x
 #define __STRINGIFY(x) __STRINGIFY_HELPER(x)
 
 [[noreturn]] void __assertion_failed(const char* msg, const char* file, unsigned line, const char* func);
-#define VERIFY(expr) (static_cast<bool>(expr) ? void(0) : __assertion_failed(#expr, __FILE__, __LINE__, __PRETTY_FUNCTION__))
+#define VERIFY(expr) (LIKELY(expr) ? void(0) : __assertion_failed(#expr, __FILE__, __LINE__, __PRETTY_FUNCTION__))
```
Comment (alimpfard) [inline Kernel/Assertions.h:15]: tbh I'd rather have `__builtin_expect` used directly here
Humans are normally pretty bad with annotating these, and giving them an easy macro tends to shit on perf _and_ readability (that's my experience from C projects, but I imagine the same applies here)

File: AK/Platform.h [hunk 2]
```diff
@@ -40,6 +40,21 @@
...
+
+#ifdef LIKELY
+#    undef LIKELY
+#endif
+#define LIKELY(expr) __builtin_expect(static_cast<bool>(expr), true)
+
+#ifdef UNLIKELY
+#    undef UNLIKELY
+#endif
+#define UNLIKELY(expr) __builtin_expect(static_cast<bool>(expr), false)
```
Comment (Hendiadyoin1) [inline AK/Platform.h:56]: there are `[[likely]]` and `[[unlikely]]` in the spec and hav support from both gcc and clang,
to which are these different than slapping one of them after the conditional?

I guess there can be more fine grained, though there arent many cases where you have mixed likability

Comment (BertalanD) [reply AK/Platform.h:56]: Unfortunately, they cannot be used with the ternary operator which we use for `VERIFY`.

As Ali suggested, I'll probably just remove these macros altogether. My initial idea was to allow making these macros into no-ops if a compiler does not support `__builtin_expect`, but I guess we won't be compiling with MSVC anytime soon, so there is no need to worry about that.

Comment (Hendiadyoin1) [reply AK/Platform.h:56]: If I am reading it correctly  they actually might,
atleast cppreference.com uses it like that, if I am not mistaken 

https://en.cppreference.com/w/cpp/language/attributes/likely

other wise can you make it an if, or do line numbers change from that?
```cpp
do {
  if (!(expr)) [[unlikely]]
    __assertion_failed(#expr, __FILE__, __LINE__, __PRETTY_FUNCTION__);
} while (0)
```

**Pull request merged**